crates/sparrow-execution/src/lib.rs

```diff
@@ -8,5 +8,228 @@
     clippy::undocumented_unsafe_blocks
 )]
 
-//! Implementations of the pipelines to be executed.
+//! Execution of physical plans.
+
+use arrow_array::RecordBatch;
+use arrow_schema::{DataType, Field, Schema, SchemaRef, TimeUnit};
+use hashbrown::HashMap;
+use sparrow_interfaces::{ReadConfig, Source};
+use sparrow_physical::StepId;
+use sparrow_transforms::TransformPipeline;
+use std::sync::Arc;
+
+use error_stack::ResultExt;
+use itertools::Itertools;
+use sparrow_scheduler::{InputHandles, WorkerPool};
+use uuid::Uuid;
+
+mod error;
+mod source_tasks;
+mod write_channel_pipeline;
+
+#[cfg(test)]
 mod tests;
+
+pub use error::*;
+
+use crate::source_tasks::SourceTasks;
+use crate::write_channel_pipeline::WriteChannelPipeline;
+
+pub struct PlanExecutor {
+    worker_pool: WorkerPool,
+    source_tasks: SourceTasks,
+    read_config: Arc<ReadConfig>,
```

Comment:
I don't think we would. How would two sources in the same query have state at different times?

`(sourceA.col(x) + sourceB.col(y)).sum()`

The aggregation state is over a combination of `sourceA` and `sourceB`.

Similarly, the options in `ReadConfig` are actually part of the `ExecutionConfig` passed in. I suspect:

a. These options should be part of `ExecutionConfig` and just passed to the plan executor in an `Arc`.
b. There may be other options that *are* read specific (eg., predicate pushdown, environment permissions, etc.)

Reply:
I'd think it depends on how we manage state. It's possible for a source to be ahead of another source, up until the `merge` channel it produces to fills up and is blocked on sending new batches. In that case, the `merge` state may hold up to time `T+10` for source `A`, but only up to `T` for source `B`. In that case, we'd want source `A` to begin reading from `T+10`, and source `B` from `T`. 

I suspect (worry) it may be more complicated than just that, though.

Reply:
Hmm... interesting point. I think we should probably discuss separately -- this would be stateful *merge* not just stateful *aggregation*. I'm not sure how well it would work out, when we consider per-entity state and/or late data. But I haven't thought much about it.

Reply:
Yeah I think we'll have to dedicate some more time to it. But at the least, any stateful operation will need to save state. Doing _just_ stateful aggregation doesn't give us anything, does it?